pkg/command/map_string.go

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 Authors of Cilium
+
+package command
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/spf13/cast"
+	"github.com/spf13/viper"
+)
+
+var keyValueRegex = regexp.MustCompile(`((\w+)=(\w+,))*((\w+)=(\w+))$`)
+
+// GetStringMapString contains one enhancement to support k1=v2,k2=v2 compared to original
+// implementation of GetStringMapString function
+// Related upstream issue https://github.com/spf13/viper/issues/911
+func GetStringMapString(vp *viper.Viper, key string) map[string]string {
+	v, _ := GetStringMapStringE(vp, key)
+	return v
+}
+
+// GetStringMapStringE is same as GetStringMapString, but with error
+func GetStringMapStringE(vp *viper.Viper, key string) (map[string]string, error) {
+	data := vp.Get(key)
+	if data == nil {
+		return map[string]string{}, nil
+	}
+	v, err := cast.ToStringMapStringE(data)
+	if err != nil {
+		var syntaxErr *json.SyntaxError
+		if !errors.As(err, &syntaxErr) {
+			return v, err
+		}
+
+		switch s := data.(type) {
+		case string:
+			if len(s) == 0 {
+				return map[string]string{}, nil
+			}
+
+			// if the input is starting with either '{' or '[', just preserve original json parsing error.
+			firstIndex := firstNonSpaceCharIndex(s)
+			if firstIndex != -1 && (s[firstIndex] == '{' || s[firstIndex] == '[') {
+				return v, err
+			}
+
+			if !isValidKeyValuePair(s) {
+				return map[string]string{}, fmt.Errorf("'%s' is not formatted as key=value,key1=value1", s)
+			}
+
+			var v = map[string]string{}
+			kvs := strings.Split(s, ",")
+			for _, kv := range kvs {
+				temp := strings.Split(kv, "=")
+				if len(temp) != 2 {
+					return map[string]string{}, fmt.Errorf("'%s' is not formatted as key=value,key1=value1", s)
+				}
+				v[temp[0]] = temp[1]
+			}
+			return v, nil
+		}
+	}
+	return v, nil
+}
+
+// isValidKeyValuePair returns true if the input is following key1=value1,key2=value2,...,keyN=valueN format.
+func isValidKeyValuePair(str string) bool {
+	if len(str) == 0 {
+		return true
+	}
+	return len(keyValueRegex.ReplaceAllString(str, "")) == 0
+}
+
+// firstNonSpaceCharIndex returns the first non-space character index
+func firstNonSpaceCharIndex(str string) int {
+	for i := range str {
+		if !isSpace(str[i]) {
+			return i
+		}
+	}
+	return -1
+}
+
+// isSpace returns true if input is space
+// this is same as the function used in json package
+func isSpace(c byte) bool {
```

Comment:
@sayboras Before this merges, can you take a look at the "unicode" package. I think some of these methods are redundant, such as: https://pkg.go.dev/unicode#IsSpace

Once over "unicode", "strings" and "bytes" packages to confirm if any of these helper methods have analogies in the stdlib.

Reply:
thanks, I was checking this one from std library, it was a private method though.

https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/encoding/json/scanner.go;l=197-199

Reply:
you are right, it's much simpler with the below

```golang
			// if the input is starting with either '{' or '[', just preserve original json parsing error.
			firstIndex := strings.IndexFunc(s, func(r rune) bool {
				return !unicode.IsSpace(r)
			})
```